### PR TITLE
[jit] Assert that the trampoline code is called in gc unsafe mode, th…

### PR DESCRIPTION
**Pull request by vargaz**
…row exceptions by returning the exception to the assembly trampoline code and throwing it from there.

### CODE REVIEW

**Comment by kumpera** [comment]
Not sure what's the goal of this change

**Comment by vargaz** [comment]
The deleted lines throw the exception using mono_raise_exception () which we are trying to get rid of. Also, icalls work like this:
- call icall
- check for pending exception
- if there is one, call another icall to retrieve it and throw it from managed code.
What happens currently is that when calling the icall which retrieves the pending exception, the trampoline code gets invoked to resolve it, and that will itself throw the exception, so this is a mess.

Also, I don't think the trampoline/jit code ever should be called in gc safe mode, which is for short predictable blocks of native code.

File: mono/utils/mono-threads-state-machine.c [hunk 1]
```diff
@@ -633,3 +633,22 @@ mono_thread_state_name (int state)
 {
 	return state_name (state);
 }
+
+gboolean
+mono_thread_is_gc_unsafe_mode (void)
+{
+	MonoThreadInfo *cur = mono_thread_info_current ();
+	int state;
```
Comment (alexrp) [inline mono/utils/mono-threads-state-machine.c]: Unnecessary variable?

Comment (vargaz) [inline mono/utils/mono-threads-state-machine.c]: which one ?

Comment (alexrp) [inline mono/utils/mono-threads-state-machine.c]: `state` is assigned below but isn't used.

**Comment by kumpera** [comment]
@vargaz thanks for the explanation, makes perfect sense.

File: mono/mini/mini-trampolines.c [hunk 2]
```diff
@@ -824,19 +824,19 @@ gpointer
 mono_magic_trampoline (mgreg_t *regs, guint8 *code, gpointer arg, guint8* tramp)
 {
 	gpointer res;
+	MonoError error;
 
-	MONO_ENTER_GC_UNSAFE_UNBALANCED;
+	MONO_REQ_GC_UNSAFE_MODE;
 
-	MonoError error;
+	g_assert (mono_thread_is_gc_unsafe_mode ());
```
Comment (luhenry) [inline mono/mini/mini-trampolines.c]: This is redundant with https://github.com/mono/mono/pull/5409/files#diff-9d04cbdf08cdb766439413f26549f9c7R829

Comment (vargaz) [reply mono/mini/mini-trampolines.c]: What it is redundant with ? The MONO_REQ_GC_UNSAFE_MODE line is only executed in checked mode.

**Pull request merged**